File: PhaoHoaDemoPublic/GRAPHICS/Csharp/PhaoHoaDemo01/Hat.cs
﻿using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;

namespace PhaoHoaDemo01
{
    // hạt sáng
    class Hat
    {
        public PointF ToaDo { get; protected set; }     // tọa độ
        public PointF VanToc { get; protected set; }    // vận tốc, đúng nghĩa thì dùng kiểu Vector sẽ tốt hơn
        public Brush CoVe { get; protected set; }       // cọ vẽ, hỗ trợ tạo màu


        public int ThoiGianSong { get; protected set; } // thời gian sống, thời gian sống <= 0 tức là đã chết


        public Hat(PointF toado, PointF vantoc, int thoigiansong, Brush cove)
        {
            this.ToaDo = toado;
            this.VanToc = vantoc;
            this.ThoiGianSong = thoigiansong;
            this.CoVe = cove;

            if (this.CoVe == null)
                this.CoVe = Brushes.Red;
        }


        public virtual void Ve(Graphics g)
        {
            g.FillRectangle(this.CoVe, ToaDo.X - 1, ToaDo.Y - 1, 2, 2);
        }


        // trả về true nếu hạt còn sống, trả về false nếu hạt đã chết (hết thời gian sống)
        public virtual bool CapNhat(Graphics g)
        {
            if (ThoiGianSong <= 0)
                return false;

            ToaDo = new PointF(ToaDo.X + VanToc.X, ToaDo.Y + VanToc.Y);

            ThoiGianSong--;
            
            Ve(g);

            return (ThoiGianSong > 0);
        }
    }
}

File: PhaoHoaDemoPublic/GRAPHICS/Java/PhaoHoaDemo01/src/phaohoademo01/Hat.java
package phaohoademo01;

import java.awt.Graphics;
import java.awt.geom.Point2D;

// hạt sáng
public class Hat {
    public Point2D.Double ToaDo;        // tọa độ
    public Point2D.Double VanToc;       // vận tốc, đúng nghĩa thì dùng kiểu Vector sẽ tốt hơn

    public int ThoiGianSong;            // thời gian sống, thời gian sống <= 0 tức là đã chết
    
    
    public Hat(Point2D.Double toado, Point2D.Double vantoc, int thoigiansong) {
        this.ToaDo = (Point2D.Double)toado.clone();
        this.VanToc = (Point2D.Double)vantoc.clone();
        this.ThoiGianSong = thoigiansong;
    }


    public void CapNhatUI(Graphics g) {
        g.fillRect((int)ToaDo.x - 1, (int)ToaDo.y - 1, 2, 2);
    }


    // trả về true nếu hạt còn sống, trả về false nếu hạt đã chết (hết thời gian sống)
    public boolean CapNhat() {
        if (ThoiGianSong <= 0)
            return false;

        ToaDo.x += VanToc.x;
        ToaDo.y += VanToc.y;

        ThoiGianSong--;

        //CapNhatUI(g);

        return (ThoiGianSong > 0);
    }
}
